Loadbalancer and NAT-Gateway Proposal (#518)
onmetal/onmetal-api: docs/proposals/03-loadbalancer.md

---
title: Network Loadbalancer

oep-number: 3

creation-date: 2022-10-18

status: implementable

authors:

- "@gehoern"
reviewers:
- "@adracus"
- "@afritzler"
- "@guvenc"
- "@MalteJ"

---

# OEP-3: Network Loadbalancer

## Table of Contents

- [Summary](#summary)
- [Motivation](#motivation)
    - [Goals](#goals)
    - [Non-Goals](#non-goals)
    - [Details](#details)
- [Proposal](#proposal)

## Summary

> Network service for Load Balancing is needed in every modern infrastructure (e.g. Kubernetes). Load balancing makes services scalable, fault-tolerant and accessible from external users. The load balancer described here is a L3-Loadbalancer (so it implements an IP level load balancing). The load balancer function is a standard network function that can be to be deployed (as an implementation detail) on a platform/node that is capable of running network functions (e.g. dp-service enabled nodes/nic's) 

Reviewer: We should omit the part about how load balancers might be deployed, especially mentioning `dp-service`.

Author: done

## Motivation

> The onMetal network is a fully routed network. To run workloads hosted on several nodes (in Kubernetes terms equivalent to cluster IP's) and not only on a single node (node IP's) solutions like ECMP (Equal cost multi pathing) seem to be the natural solution. But ECMP has certain limitations: the amount of nodes it can be used to distribute traffic to and a hash based algorithm to pin the network flows, that changes - based on the amount of targets - its distributed pinning. 

Reviewer: Please replace `onMetal` with `onmetal` (applies for all occurrences)

> The onMetal network is a fully routed network. To run workloads hosted on several nodes (in Kubernetes terms equivalent to cluster IP's) and not only on a single node (node IP's) solutions like ECMP (Equal cost multi pathing) seem to be the natural solution. But ECMP has certain limitations: the amount of nodes it can be used to distribute traffic to and a hash based algorithm to pin the network flows, that changes - based on the amount of targets - its distributed pinning. 

Reviewer: The motivation here goes into the direction of how an LB can be implemented. In `onmetal-api` we should rather focus on the general motivation why Load Balancers are necessary. In our case e.g., we have the `VirtualIP` which is a single stable public IP attached to a `NetworkInterface`, however, to scale beyond single network interfaces, `LoadBalancer`s offer the capability to dynamically route traffic to multiple targets.

Reviewer 2: I believe, yeah, considerations regarding exact implementation should be moved to the project that will handle the provisioning of load balancer service, e.g. `lblet`.

Also, it is not clear from the motivation whether ECMP will be used or not, which features would be included (e.g. will L3DSR be supported), etc.

Author: tried to make it less specific but still describing the basic idea. Took your sentence @adracus

To avoid those limitations and to have better control over the load balancer behavior an onMetal load balancer service needs to be provided.

### Goals

> - define a user facing API (onMetal api) for the load balancer

Reviewer 3: Please start all lists with an upper case.

Author: done

> - enable the user to request an L3 Loadbalancer (only IP addresses) with load balancer address (outside facing) and 

Reviewer: Meld into sth like `Define an API for managing L3 load balancers with publicly available addresses`

Author: done

- with multiple target's based on the onMetal ([OEP-1](01-networking-integration.md)) network interface architecture

> - the targets of the load balancer must be dynamic changeable

Reviewer: ```suggestion
- the targets of the load balancer should be dynamically changeable
```

- the load balancer needs to avoid forwarding unintended traffic.
  - based on ports/protocols (UDP/TCP/SCTP) it must filter what needs to be load balanced

>   - ICMP requests are not part of forwarded traffic and will be filtered by the load balancer anyhow (the loa)

Reviewer: What about

> (the loa)

?

Author: typo

- Load balancer (IP) and targets can be parts of different networks but

> - all targets (NetworkInterfaces) must be of the same network 

Reviewer: ```suggestion
- all targets (`NetworkInterface`s) must be in the same `Network` 
```

Author: done

### Non-Goals
- no address or port translation / rewriting (no SNAT / DNAT) (L4 Loadbalancer) support
- no injection of additional information (e.g. x-forwarded-for) (L7 Loadbalancer) support
- no protocol offloading like ssl (L7 Loadbalancer) support
- if multiple load balancer IP's are needed, request multiple load balancers (a load balancer services one load balancer IP)

### Details
- load balancing is used to deliver a packet addressed to the load balancer to one of its targets via the onMetal network routing

> - the load balancing will be for the target and the source transparent

Reviewer: ```suggestion
- the load balancing will transparent for both the target and the source
```

Author: yes nicer

> - the target needs to know the load balancers IP and needs to answer with it (and to receive traffic with it)

Reviewer: ```suggestion
- the target needs to know the load balancer's IP and needs to answer with it (and to receive traffic with it)
```

Author: done

- answers to request will be directly delivered since all details are known to the target
- payload packages are not changed, to not lose any information

## Proposal

> A network loadbalancer CRD allows the user to define the network function. all entries except the `targetNetworkInterface` must be immutable. 

Reviewer: ```suggestion
A `LoadBalancer` allows the user to define the network function. `type`, `ipFamilies` and `networkRef` are immutable.
```
Regarding immutability, potentially also `ports` should be mutable.

Author: written different, since in the last iteration everything is immutable but the networkInterfaceSelector allows for dynamic targets.

> ### User defined object

Reviewer 2: In most clouds it is possible to define and request a load balancing objects as `Service` resources with certain types like `LoadBalancer` or `NodePort` (in case of MetalLB), and with `Ingress` resource.

Probably it is worth to mention how these objects will correlate as it would allow to avoid any further compatibility and implementation issues.

Reviewer: Kubernetes `Service`s of type `NodePort` are a feature of the `Node`s running in a Cluster.
`LoadBalancer`s are inherently cloud provider specific.

Author: I mentioned explicitly the usage of a Kubernetes ingess. hope that helpes.

```yaml
apiVersion: networking.api.onmetal.de/v1alpha1
kind: LoadBalancer
metadata: 
  name: myLoadBalancer-2abf34
  namespace: customer-1
spec:
  #the load balancer IP. is a generated IP and only of type public (internet IP) 
  #TODO other types of IP addresses (e.g. internal LB) and pinned IP types
  type: Public
  #a load balancer can have two IP addresses: IPv4 or IPv6 but not more addresses IPfamily
  ipFamilies: [ IPv4, IPv6 ]

  #filtering the load balanced traffic by protocol and port/port ranges 
  ports:
    - name: webserver  # might be optional
      # protocols supported UDP, TCP, SCTP
      protocol: tcp
      # single port
      port: 80
    - name: db
      protocol: udp
      # portrange

>       port: 1024

Reviewer 2: I'd perefer to have for port range

```yaml
portRange: 
  begin: 10000
  end: 11000
```

Reviewer: We try to integrate into the k8s environment as naturally as possible and took this approach from how k8s does network policies, see
https://kubernetes.io/docs/concepts/services-networking/network-policies/#targeting-a-range-of-ports
As we already strive to be as kubernetes-like as possible, we should continue here.

Author: 👍 @adracus that is exactly the source of the definition. btw. the implicit advantage here is to have a dedicated port and a port range in the same easy definition.

      portEnd: 2048
  #all networkInterfaces must be connected to this network.
  networkRef:
    name:
  #a selector for the NetworkInterfaces to be load balancer targets, normal kubernetes selector logic

>   #TODO explicit machines ?

Reviewer: Please check / remove all TODOs.

  networkInterfaceSelector:
    matchLabels: 
      key: db
      foo: bar
status:

>   #the ips the load balancer is listening to (max 2, an IPv4 and/or IPv6)

Reviewer 2: For extensible scalability it'd be good to support multiple IPs per load balancer, i.e. supply and configure them with same set of rules.

Then, it'd be possible to put them into the RRDNS and have multiple resolvable IP addresses per endpoint.

Reviewer: That's exactly the non-goal we mentioned - if multiple IPs should be supported, multiple LBs have to be created.

Author: in theory that is a nice idea, but the idea so so nice scalable that there should be no technical reason for having multiple ip addresses (e.g. performance distribution is not an issue here) and therefore we explicitly define a single LB for a single purpose. If you want to have the same LB semantics but with a different LB IP, you can have this by defining a copy of the LB object, but you can also realise as a customer that this may not be intended

  ips: 
    - 45.86.152.88
    - 2001::
```

> ### internal reconcile object (not customer visible)

Reviewer: We don't care about customer visibility here - that's an aspect that will be handled via RBAC.
The `LoadBalancerRouting` resource is more a resource that shows the expanded routing state at a certain point in time, hence the wording `internal` does not quite match.

Author: fixed

```yaml
apiVersion: networking.onmetal.de
kind: LoadBalancer
metadata: 
  name: myLoadBalancer-2abf34
  namespace: customer-1
spec:
  type: Public
  ipFamilies: [ IPv4, IPv6 ]

  # limits the networkinterfaces to select from
  networkRef:
    name: 
  ports:
    - name: webserver  # might be optional

>       protocol: tcp

Reviewer: The LB is L3. Filtering will be L4 and supports multiple protocols at the same time.

>       protocol: tcp

Author: 👍 @adracus, clarified in text

```

```yaml
apiVersion: networking.api.onmetal.de/v1alpha1

> kind: LoadBalancerRouting

Reviewer 4: Which object should be reconciled **LoadBalancer** or **LoadBalancerRouting**?

Reviewer: The `onmetal-api` controllers will apply `LoadBalancerRouting` from `LoadBalancer`.

Author: hope i clarified the text, as @adracus did here
